### ISSUE TITLE
ilsresc -l prints CAT_NO_ROWS_FOUND with exactly 350 resources, with return code 0

### ISSUE THREAD

**Issue by bh9** [irods/irods#5155]
- [x] master
- [x] 4-2-stable
-----

### iRODS Version, OS and Version
4.2.7 Ubuntu 18.04 running 16.04 packages

### What did you try to do?
`ilsresc -l` in a zone with 350 resources

### Expected behavior
`ilsresc -l` outputs the long form information about every resource in the zone and sends nothing to stderr

### Observed behavior (including steps to reproduce, if applicable)
`ilsresc -l` outputs the long form information about every resource in the zone and sends `rcGenQuery failed with error -808000 CAT_NO_ROWS_FOUND` to stderr

**Comment by trel** [comment]
Can you confirm that all the resources expected are in the output of `ilsresc -l`?  If not, only one missing?  More than one?

The `CAT_NO_ROWS_FOUND` is because a resource name lookup didn't succeed?  Non-standard character in the name?  Truncated serialization between two pieces of this command?  Anything useful in the rodsLog?

**Comment by bh9** [comment]
There's one resource not present in the list, but that resource is `bundleResc` which is also not present in the list when there's fewer than 350 resources

**Comment by bh9** [comment]
Nothing in the rodslog. The resource that tips it over is no different to the 4 before it, so it's nothing special about that resource. This one is fine:
```
irods@irods-seq-sb01:~$ iadmin mkresc irods-seq-sb05-ddn-bd12b-vd0-4 unixfilesystem irods-seq-sb05.internal.sanger.ac.uk:/irods-seq-sb05-ddn-bd12b-vd0-4
Creating resource:
Name:           "irods-seq-sb05-ddn-bd12b-vd0-4"
Type:           "unixfilesystem"
Host:           "irods-seq-sb05.internal.sanger.ac.uk"
Path:           "/irods-seq-sb05-ddn-bd12b-vd0-4"
Context:        ""
irods@irods-seq-sb01:~$ ilsresc -l >/dev/null
```
and then this one is not:
```
irods@irods-seq-sb01:~$ iadmin mkresc irods-seq-sb05-ddn-bd12b-vd0-5 unixfilesystem irods-seq-sb05.internal.sanger.ac.uk:/irods-seq-sb05-ddn-bd12b-vd0-5
Creating resource:
Name:           "irods-seq-sb05-ddn-bd12b-vd0-5"
Type:           "unixfilesystem"
Host:           "irods-seq-sb05.internal.sanger.ac.uk"
Path:           "/irods-seq-sb05-ddn-bd12b-vd0-5"
Context:        ""
irods@irods-seq-sb01:~$ ilsresc -l >/dev/null
rcGenQuery failed with error -808000 CAT_NO_ROWS_FOUND
```

**Comment by trel** [comment]
Strange.   We'll probably refactor `ilsresc` to use the new query iterator - and these types of edge cases should already be handled.

**Comment by kript** [comment]
I'm a bit twitchy that until we understand why this doesn't work we will hit a similar restriction in something we can't work around, so I would be very interested in knowing why `ilsresc` goes 💥 on that no of resources  as well as potentialy refactoring it away please.
Just in case the issue was going to sit on backlog until you get a chance to refactor!

**Comment by trel** [comment]
My initial effort to reproduce this issue did not succeed.   More eyeballs coming.

**Comment by trel** [comment]
I would hop not, but ... any difference if the `ilsresc -l` is issued on provider vs consumer?

**Comment by kript** [comment]
We only run this on consumers as providers don't have resources (the initial example is a consumer), and we get a lot of log spam and potential loss of access when we tip over the limit so not keep to put back and find out, sorry! 🤷‍♂️

**Comment by korydraughn** [comment]
I haven't been able to reproduce this issue either. I created 400 ufs resources and 4 replication resources. The results were the same for the consumer and provider.

**Comment by bh9** [comment]
Update on this after further testing: when iadmin lr shows _exactly_ 351 resources, ilsresc shows 350 resources and throws the error. At 350 and 352 in iadmin lr, there is no error. That would suggest to me that it's likely a pagination issue, where ilsresc is looking for another page, but because bundleResc is excluded, that page empty so we get CAT_NO_ROWS_FOUND

**Comment by trel** [comment]
Behold!

```
$ iadmin lr | wc -l
351
$ ilsresc | wc -l
350
$ ilsresc -l >/dev/null
rcGenQuery failed with error -808000 CAT_NO_ROWS_FOUND 
$ echo $?
0
```

Agreed, this is just a boundary issue and a query iterator should clean it right up.

**Comment by kript** [comment]
This also happens at other boundary issues;

```
$ iadmin lr  | wc -l
201
$ ilsresc -l > /tmp/ilsresc
rcGenQuery failed with error -808000 CAT_NO_ROWS_FOUND
```

**Comment by kript** [comment]
```
$ iadmin mkresc workaroundfor5155 passthru
Creating resource:
Name:           "workaroundfor5155"
Type:           "passthru"
Host:           ""
Path:           ""
Context:        ""
$ iadmin lr  | wc -l
202
$ ilsresc -l > /tmp/ilsresc
$
```

worked around by dummy resource increase

**Comment by SwooshyCueb** [comment]
I'm in favor of pushing this to 4.2.10 or the 4.2 backlog if nobody else thinks they can get to it before the 4.2.9 deadline.

**Comment by trel** [comment]
Agreed - a refactor shouldn't be too hard, but we are out of time for this release.

**Comment by korydraughn** [comment]
refactoring `ilsresc` to use the query iterator is one way to resolve this, but it is unnecessary.

this error message occurs when the GenQuery resultset is a multiple of the page size (which is hard coded to 50 (see [here](https://github.com/irods/irods_client_icommands/blob/37ba4006b86933dfdef8e16f5eddc44a6962f8b6/src/ilsresc.cpp#L160)).

this is resolved by explicitly checking for `CAT_NO_ROWS_FOUND` near the [location where the error message is printed](https://github.com/irods/irods_client_icommands/blob/37ba4006b86933dfdef8e16f5eddc44a6962f8b6/src/ilsresc.cpp#L44-L46).

this is fixed by the following PR: https://github.com/irods/irods_client_icommands/pull/236

**Comment by trel** [comment]
checkboxed.

**Comment by kript** [comment]
Just it this with 251 resources on a different zone - is it in the fix list for 4.2.11?

**Comment by trel** [comment]
Yes, this is fixed/closed and will be in 4.2.11.
